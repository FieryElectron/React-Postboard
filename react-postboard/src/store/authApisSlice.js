import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

export const getAuthApis = createAsyncThunk(
    "authApis/getAuthApis",
    async () => {
        return await fetch(`http://` + process.env.REACT_APP_ROOT_DOMAIN + `:` + process.env.REACT_APP_AUTH_PORT + `/api/`)
            .then((res) => res.json());
    }
);

const authApisSlice = createSlice({
    name: "authApis",
    initialState: {
        apis: {},
        status: null,
    },
    extraReducers: {
        [getAuthApis.pending]: (state, action) => {
            state.status = "loading";
        },
        [getAuthApis.fulfilled]: (state, action) => {
            state.status = "success";
            state.apis = action.payload;
        },
        [getAuthApis.rejected]: (state, action) => {
            state.status = "failed";
        },
    },
});

export default authApisSlice.reducer;