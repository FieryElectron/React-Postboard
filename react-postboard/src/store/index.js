import { configureStore } from "@reduxjs/toolkit";
import utilReducer from "./utilSlice";
import authApisReducer from "./authApisSlice";
import restApisReducer from "./restApisSlice";
import userReducer from "./userSlice";
import postsReducer from "./postsSlice";

const store = configureStore({
    reducer: {
        authApis: authApisReducer,
        restApis: restApisReducer,
        util: utilReducer,
        user: userReducer,
        posts: postsReducer,
    }
})

export default store;