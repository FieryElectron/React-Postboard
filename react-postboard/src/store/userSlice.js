import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { useSelector } from 'react-redux';
import axios from 'axios';
import { useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';

const refreshAccessToken = async (tokenUrl) => {
    const res = await axios.get(tokenUrl, {withCredentials: true});
    return res.data.flag;
}

const tryLoadProfile = async (loadprofileUrl, tokenUrl, navigate, reTry) => {
    console.log("tryLoadProfile");

    const res = await axios.get(loadprofileUrl, {withCredentials: true});
    
    if(res.data.flag){
        return res.data.user;
    }

    if(reTry){
        reTry = false;
        if(await refreshAccessToken(tokenUrl)){
            console.log("Access Token Refreshed!");
            return await tryLoadProfile(loadprofileUrl, tokenUrl, navigate , reTry);
        }else{

            // toast.info("Back to Sign In Page!", {
            //     position: "top-center",
            //     autoClose: 1000,
            //     pauseOnFocusLoss:false,
            // });
            
            navigate("/");

            return {username:"",description:""};
        }
    }
}

export const getUser = createAsyncThunk(
    "user/getUser",
    async ({navigate, restApis}, thunkAPI) => {
        const loadprofileUrl = restApis.loadprofile;
        const tokenUrl = restApis.token;

        return await tryLoadProfile(loadprofileUrl, tokenUrl, navigate, true);
    }
); 
 
const userSlice = createSlice({
    name: "user",
    initialState: {
        user: {username:"",description:""},
        status: null,
    },
    extraReducers: {
        [getUser.pending]: (state, action) => {
            state.status = "loading";
        },
        [getUser.fulfilled]: (state, action) => {
            state.status = "success";
            state.user = action.payload;
        },
        [getUser.rejected]: (state, action) => {
            state.status = "failed";
        },
    },
});

export default userSlice.reducer;