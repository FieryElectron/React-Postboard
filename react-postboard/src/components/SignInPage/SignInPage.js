import Header from '../Header/Header';
import Button from '../Button/Button';

import Input from '../Input/Input';
import { useNavigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

import { utilActions } from '../../store/utilSlice';

import { toast } from 'react-toastify';

import axios from 'axios';
import { useEffect } from 'react';
import { getAuthApis } from '../../store/authApisSlice';

import { GoogleLogin } from 'react-google-login';

import { gapi } from 'gapi-script';

import GoogleSignInButton from '../GoogleSignInButton/GoogleSignInButton';

const clientId = process.env.REACT_APP_CLIENT_ID;

const SignInPage = () => {
    const dispatch = useDispatch();

    useEffect(() => {
      dispatch(getAuthApis());
      

      function start(){
        gapi.client.init({
            clientId:clientId,
            scope:"profile",
        })
      }

      gapi.load('client:auth2', start);
    }, [dispatch]); 

    const navigate = useNavigate();

    const username = useSelector((state) => state.util.username);
    const password = useSelector((state) => state.util.password);

    const registerUrl = useSelector((state) => state.authApis.apis.register);
    const loginUrl = useSelector((state) => state.authApis.apis.login);
   

    const usernameOnChange = (e) => {
        dispatch(utilActions.setUsername(e.target.value));
    }

    const passwordOnChange = (e) => {
        dispatch(utilActions.setPassword(e.target.value));
    }

    const signUp = async () => {
        const url = registerUrl+`?username=${username}&password=${password}`;
        const res = await axios.post(url);

        toast.info(res.data.info, {
            position: "top-center",
            autoClose: 1000,
            pauseOnFocusLoss:false,
        });
    }

    const signIn = async () => {
        const url = loginUrl+`?username=${username}&password=${password}`;
        const res = await axios.post(url, {}, {withCredentials: true});

        toast.info(res.data.info, {
            position: "top-center",
            autoClose: 1000,
            pauseOnFocusLoss:false,
        });
        
        if(res.data.flag){
            navigate("/postboard");
            dispatch(utilActions.setUsername(""));
            dispatch(utilActions.setPassword(""));
        }
    }

    const loginGoogle = async () => {

    }

    return (
        <div>
            <Header title='Sign In' />
            <Input placeholder='Username' onChange={usernameOnChange}/>
            <Input placeholder='Password' type='password' onChange={passwordOnChange}/>
            <br />
            <Button text='Sign Up' onClick={signUp} />
            <Button text='Sign In' onClick={signIn} />
            <br />
            <GoogleSignInButton/>

        </div>
    )
}

export default SignInPage;