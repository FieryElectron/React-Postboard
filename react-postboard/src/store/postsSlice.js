import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { useSelector, useDispatch } from 'react-redux';
import axios from 'axios';
import { toast } from 'react-toastify';
import { utilActions } from './utilSlice';
import $ from 'jquery';





const refreshAccessToken = async (tokenUrl) => {
    const res = await axios.get(tokenUrl, {withCredentials: true});
    return res.data.flag;
}

const tryPostPagination = async (thunkAPI, navigate, postUrl, tokenUrl, reTry) => {

    const page = thunkAPI.getState().util.selectedPage;
    const fromDate = thunkAPI.getState().util.fromDate;
    const toDate = thunkAPI.getState().util.toDate;
    

    console.log("tryPostPagination",page);

    const from = new Date(fromDate).getTime();
    const to = new Date(toDate).getTime();

    const res = await axios.get(postUrl + `?from=` + from + `&to=` + to + `&page=` +page, {withCredentials: true});

    if(res.data.flag){
        // console.log(res.data.pages,page);

        let withoutContentPage = false;

        if(res.data.pages < page && res.data.pages > 0){
            withoutContentPage = true;
        }
        let pages = [];
        for(let i=0;i<res.data.pages;++i){
            pages.push(i+1);
        }
        thunkAPI.dispatch(utilActions.setPages(pages));

        if(withoutContentPage){
            const lastPage = pages.at(-1);
            thunkAPI.dispatch(utilActions.setSelectedPage(lastPage));
            $("#pageSelectorId").val(lastPage);

            return await tryPostPagination(thunkAPI, navigate, postUrl, tokenUrl, true);
        }

        return res.data.posts;
    }

    if(reTry){
        reTry = false;
        if(await refreshAccessToken(tokenUrl)){
            console.log("Access Token Refreshed!");
            return await tryPostPagination(thunkAPI, navigate, postUrl, tokenUrl, reTry);
        }else{
            // toast.info("Back to Sign In Page!", {
            //     position: "top-center",
            //     autoClose: 1000,
            //     pauseOnFocusLoss:false,
            // });
            
            navigate("/");

            return [];
        }
    }
}

export const getPosts = createAsyncThunk(
    "posts/getPosts",
    async ({navigate, restApis}, thunkAPI) => {
        const postUrl = restApis.post;
        const tokenUrl = restApis.token;

        return await tryPostPagination(thunkAPI, navigate, postUrl, tokenUrl, true);
    }
); 
  
const postsSlice = createSlice({
    name: "posts",
    initialState: {
        posts: [],
        status: null,
    },
    extraReducers: {
        [getPosts.pending]: (state, action) => {
            state.status = "loading";
        },
        [getPosts.fulfilled]: (state, action) => {
            state.status = "success";
            state.posts = action.payload;
        },
        [getPosts.rejected]: (state, action) => {
            state.status = "failed";
        },
    },
});

export default postsSlice.reducer;