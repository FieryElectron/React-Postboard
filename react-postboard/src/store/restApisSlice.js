import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import { getUser } from './userSlice';
import { getPosts } from './postsSlice';


export const getRestApis = createAsyncThunk(
    "restApis/getRestApis",
    async (navigate, thunkAPI) => {

        const restApis = await fetch(`http://` + process.env.REACT_APP_ROOT_DOMAIN + `:` + process.env.REACT_APP_REST_PORT + `/api/`).then((res) => res.json());

        thunkAPI.dispatch(getUser({navigate, restApis}));
        thunkAPI.dispatch(getPosts({navigate, restApis}));

        return restApis;
    }
);



const restApisSlice = createSlice({
    name: "restApis",
    initialState: {
        apis: {},
        status: null,
    },
    extraReducers: {
        [getRestApis.pending]: (state, action) => {
            state.status = "loading";
        },
        [getRestApis.fulfilled]: (state, action) => {
            state.status = "success";
            state.apis = action.payload;
        },
        [getRestApis.rejected]: (state, action) => {
            state.status = "failed";
        },
    },
});

export default restApisSlice.reducer;