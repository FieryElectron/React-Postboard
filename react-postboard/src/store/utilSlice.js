import { createSlice } from "@reduxjs/toolkit";

import { getPosts } from "./postsSlice";

import { useDispatch } from "react";
import { useNavigate } from 'react-router-dom';

const utilSlice = createSlice({
    name: 'util',
    initialState: {
        username: '',
        password: '',
        selectedPage: 1,
        fromDate: '',
        toDate: '',
        pages:[1,2,3,4,5],
        postContent: '',
    },
    reducers: {
        setUsername(state, action){
            state.username = action.payload;
        },
        setPassword(state, action){
            state.password = action.payload;
        },
        setSelectedPage(state, action){
            state.selectedPage = action.payload;
        },
        setFromDate(state, action){
            state.fromDate = action.payload;
        },
        setToDate(state, action){
            state.toDate = action.payload;
        },
        setPages(state, action){
            state.pages = action.payload;
        },
        setPostContent(state, action){
            state.postContent = action.payload;
        },
    }
})

export const utilActions = utilSlice.actions;
export default utilSlice.reducer;