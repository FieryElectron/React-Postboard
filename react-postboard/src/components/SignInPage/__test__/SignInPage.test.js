import ReactDOM from "react-dom/client";
import { render, fireEvent  } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter as Router} from 'react-router-dom'

import SignInPage from '../SignInPage';
import { Provider } from 'react-redux';
import store from '../../../store';


let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it("renders signInPage without crashing", ()=>{
    act(() => {
        ReactDOM.createRoot(container).render(<Provider store={store}><Router><SignInPage/></Router></Provider>);
    });
});

it("matchs snapshot 1", ()=>{
  const tree = renderer.create(<Provider store={store}><Router><SignInPage/></Router></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});