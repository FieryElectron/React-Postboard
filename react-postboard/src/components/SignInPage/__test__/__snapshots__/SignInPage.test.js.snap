// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchs snapshot 1 1`] = `
<div>
  <div
    className="headerContainer"
    data-testid="headerContainer"
  >
    <h1
      className="header"
      data-testid="header"
    >
      Sign In
    </h1>
  </div>
  <div
    className="inputContainer"
    data-testid="inputContainer"
  >
    <input
      className="input"
      data-testid="input"
      onChange={[Function]}
      placeholder="Username"
      spellCheck="false"
      type="text"
    />
  </div>
  <div
    className="inputContainer"
    data-testid="inputContainer"
  >
    <input
      className="input"
      data-testid="input"
      onChange={[Function]}
      placeholder="Password"
      spellCheck="false"
      type="password"
    />
  </div>
  <br />
  <button
    className="button"
    data-testid="button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#0071FE",
      }
    }
  >
    Sign Up
  </button>
  <button
    className="button"
    data-testid="button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#0071FE",
      }
    }
  >
    Sign In
  </button>
  <br />
  <button
    className="button"
    data-testid="button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#0071FE",
      }
    }
  >
    Google
  </button>
</div>
`;
